sagemaker/24_train_bloom_peft_lora/sagemaker-notebook.ipynb

```diff
@@ -201,7 +196,7 @@
     "lm_dataset = dataset.map(\n",
     "    lambda sample: tokenizer(sample[\"text\"]), batched=True, remove_columns=list(dataset.features)\n",
     ").map(\n",
-    "    partial(chunk, chunk_length=2048),\n",
```

Comment:
can we try decreasing the batch size. It is better to train on the full sequence length

Reply:
or pinning the `peft` version to `0.2.0`

Reply:
I tried a few runs pinning to peft 0.2.0 and received the following error after ~5 hours of training:

ErrorMessage "│   1908 │   │   │   raise ValueError("training mode is expected to be boolean │
 │   1909 │   │   self.training = mode                                          │
 │   1910 │   │   for module in self.children():                                │
 │ ❱ 1911 │   │   │   module.train(mode)                                        │
 │   1912 │   │   return self                                                   │
 │   1913 │                                                                     │
 │   1914 │   def eval(self: T) -> T:                                           │
 │                                                                              │
 │ /opt/conda/lib/python3.9/site-packages/torch/nn/modules/module.py:1911 in    │
 │ train                                                                        │
 │ /opt/conda/lib/python3.9/site-packages/peft/tuners/lora.py:417 in train      │
 │   414 │   │   if not mode and self.merge_weights and not self.merged:        │
 │   415 │   │   │   # Merge the weights and mark it                            │
 │   416 │   │   │   if self.r > 0 and any(self.enable_lora):                   │
 │ ❱ 417 │   │   │   │   delta_w = F.conv1d(                                    │
 │   418 │   │   │   │   │   self.lora_A.weight.data.unsqueeze(0),              │
 │   419 │   │   │   │   │   self.lora_B.weight.data.unsqueeze(-1),             │
 │   420 │   │   │   │   │   groups=sum(self.enable_lora),                      │
 ╰──────────────────────────────────────────────────────────────────────────────╯
 RuntimeError: Expected 4-dimensional input for 4-dimensional weight [8192, 8, 1,
 1], but got 3-dimensional input of size [1, 16, 4096] instead"

Reply:
'per_device_train_batch_size': 1 so I don't believe lowering the batch size is an option.

Reply:
Other items tried just to have the full debugging trail here:
- Pinning to v0.3.0 stable
- Setting os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "max_split_size_mb:X" at top of run_clm.py script with various values 516, 1500

Continuing to get GPU memory error -
ErrorMessage "OutOfMemoryError: CUDA out of memory. Tried to allocate 1.91 GiB (GPU 0; 22.19
GiB total capacity; 15.37 GiB already allocated; 1.62 GiB free; 19.37 GiB
reserved in total by PyTorch)

Reducing the dataset chunk_length was sufficient to address the GPU OOM issues. I hear you on training for the full sequence length, will continue to explore workarounds.

Reply:
I tested with peft 0.2.0 it works. Can you confirm? i would pin then.

Reply:
Thanks for testing out @philschmid. I think I was encountering the weird errors above due to testing artifacts from tinkering but I was able to confirm pinning to peft `v0.2.0` solved the OOM issue.